libcache/cache.c

```diff
@@ -0,0 +1,391 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Cache library
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Malgorzata Wrobel
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cache.h"
+
+#include <sys/list.h>
+// #include "list.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <errno.h>
+#include <string.h>
+
+#define LIBCACHE_NUM_WAYS   4
+#define LIBCACHE_ADDR_WIDTH 64
+
+
+typedef struct cacheline_s cacheline_t;
+
+
+struct cacheline_s {
+	uint64_t tag;
+	uint32_t *data;
+	unsigned char isValid;
+	unsigned char isDirty;
+	cacheline_t *prev, *next; /* Circular doubly linked list */
+};
+
+
+typedef struct cacheset_s {
+	cacheline_t *timestamps;
+	cacheline_t **tags;
+	cacheline_t lines[LIBCACHE_NUM_WAYS];
+	size_t count;
+} cacheset_t;
+
+
+struct cachectx_s {
+	cacheset_t *sets;
+
+	size_t numLines;
+	size_t numSets;
+
+	uint64_t tagMask;
+	uint64_t setMask;
+	uint64_t offMask;
+
+	off_t offBitNum;
+	off_t setBitsNum;
+	off_t tagBitsNum;
+
+	cache_readCb_t readCb;
+	cache_writeCb_t writeCb;
+};
+
+
+cacheset_t *cache_setInit(void)
+{
+	cacheset_t *set = malloc(sizeof(cacheset_t));
+
+	if (set == NULL) {
+		return NULL;
+	}
+
+	set->timestamps = NULL;
+	set->tags = calloc(LIBCACHE_NUM_WAYS, sizeof(cacheline_t *));
+	memset(set->lines, 0, sizeof(set->lines));
+
+	if (set->tags == NULL) {
+		free(set);
+		return NULL;
+	}
+
+	return set;
+}
+
+
+/* Generates mask of type uint64_t with numBits set to 1 */
+uint64_t cache_genMask(int numBits)
+{
+	return ((uint64_t)1 << (uint64_t)numBits) - (uint64_t)1;
```

Comment:
No need to cast numBits, shift-by-value needs to be in range 0-63
```suggestion
	return ((uint64_t)1 << numBits) - (uint64_t)1;
```

Reply:
Thank you, applied